File: laravel-backend/app/DataSources/Polymarket.php
<?php

namespace App\DataSources;

use Illuminate\Support\Facades\Http;

class Polymarket
{
    private const GAMMA_BASE_URL = 'https://gamma-api.polymarket.com';

    public function fetchEvents(int $limit = 20): array
    {
        try {
            $response = Http::timeout(15)->get(self::GAMMA_BASE_URL . '/events', [
                'limit' => $limit,
                'active' => 'true',
                'closed' => 'false',
            ]);

            if (! $response->successful()) {
                return [];
            }

            $events = $response->json();
            $results = [];

            foreach ($events as $event) {
                $markets = $event['markets'] ?? [];
                foreach ($markets as $market) {
                    $outcomePrices = $market['outcomePrices'] ?? '';
                    $probability = null;

                    if ($outcomePrices) {
                        $prices = json_decode($outcomePrices, true);
                        if (! empty($prices)) {
                            $probability = round((float) $prices[0] * 100, 1);
                        }
                    }

                    $results[] = [
                        'id' => $market['id'] ?? null,
                        'title' => $market['question'] ?? $event['title'] ?? null,
                        'description' => $event['description'] ?? '',
                        'probability' => $probability,
                        'volume' => $market['volume'] ?? null,
                        'end_date' => $market['endDate'] ?? null,
                    ];
                }
            }

            return $results;
        } catch (\Throwable $e) {
            return [];
        }
    }

    public function searchMarkets(string $query, int $limit = 10): array
    {
        try {
            $response = Http::timeout(15)->get(self::GAMMA_BASE_URL . '/markets', [
                'tag' => $query,
                'limit' => $limit,
                'active' => 'true',
                'closed' => 'false',
            ]);

            if (! $response->successful()) {
                return [];
            }

            return array_map(fn($m) => [
                'id' => $m['id'] ?? null,
                'title' => $m['question'] ?? null,
                'probability' => null,
                'volume' => $m['volume'] ?? null,
            ], $response->json());
        } catch (\Throwable $e) {
            return [];
        }
    }
}

File: backend/app/data_sources/polymarket.py
import httpx

GAMMA_BASE_URL = "https://gamma-api.polymarket.com"


async def fetch_events(limit: int = 20) -> list[dict]:
    """Fetch active prediction market events from Polymarket."""
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            response = await client.get(
                f"{GAMMA_BASE_URL}/events",
                params={"limit": limit, "active": "true", "closed": "false"},
            )
            response.raise_for_status()
            events = response.json()

            results = []
            for event in events:
                markets = event.get("markets", [])
                for market in markets:
                    outcome_prices = market.get("outcomePrices", "")
                    probability = None
                    if outcome_prices:
                        try:
                            # outcomePrices is a JSON string like "[0.65, 0.35]"
                            import json
                            prices = json.loads(outcome_prices)
                            if prices:
                                probability = round(float(prices[0]) * 100, 1)
                        except (json.JSONDecodeError, IndexError, ValueError):
                            pass

                    results.append({
                        "id": market.get("id"),
                        "title": market.get("question") or event.get("title"),
                        "description": event.get("description", ""),
                        "probability": probability,
                        "volume": market.get("volume"),
                        "end_date": market.get("endDate"),
                    })

            return results
    except Exception:
        return []


async def search_markets(query: str, limit: int = 10) -> list[dict]:
    """Search for prediction markets relevant to a query."""
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            response = await client.get(
                f"{GAMMA_BASE_URL}/markets",
                params={"tag": query, "limit": limit, "active": "true", "closed": "false"},
            )
            response.raise_for_status()
            markets = response.json()

            return [
                {
                    "id": m.get("id"),
                    "title": m.get("question"),
                    "probability": None,
                    "volume": m.get("volume"),
                }
                for m in markets
            ]
    except Exception:
        return []
